# Segmenting and Clustering of Data Science

## 1. Start by creating a new Notebook for this assignment.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: - 
  - anaconda/osx-64::ca-certificates-2019.8.28-0, anaconda/osx-64::certifi-2019.9.11-py37_0, anaconda/osx-64::openssl-1.1.1d-h1de35cc_2
  - anaconda/osx-64::ca-certificates-2019.8.28-0, anaconda/osx-64::certifi-2019.9.11-py37_0, defaults/osx-64::openssl-1.1.1d-h1de35cc_2
  - anaconda/osx-64::certifi-2019.9.11-py37_0, anaconda/osx-64::openssl-1.1.1d-h1de35cc_2, defaults/osx-64::ca-certificates-2019.8.28-0
  - anaconda/osx-64::certifi-2019.9.11-py37_0, defaults/osx-64::ca-certificates-2019.8.28-0, defaults/osx-64::openssl-1.1.1d-h1de35cc_2
  - anaconda/osx-64::openssl-1.1.1d-h1de35cc_2, defaults/osx-64::ca-certificates-2019.8.28-0, defaults/osx-64::certifi-2019.9.11-py37_0
  - defaults/osx-64::ca-certificates-2019.8.28-0, defaults/osx-64::certifi-2019.9.11-py37_0, defaults/osx-64::openssl-1.1.1d-h1de35cc_2
  - anaconda/osx-64::ca-certificates-2019.8.28-0, anaconda/osx-64::openssl-1.1.1d-h1de35cc_2, defaults/osx-64::certifi-2019.9.11-py37_0
  - anaconda/osx-64::ca-

# 2. transform the data into a pandas dataframe

In [2]:
# import the library we use to open URLs
import urllib.request

In [3]:
# specify which URL/web page we are going to be scraping
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [4]:
# open the url using urllib.request and put the HTML into the page variable
page = urllib.request.urlopen(url)

In [5]:
# import the BeautifulSoup library so we can parse HTML and XML documents
from bs4 import BeautifulSoup

In [6]:
# parse the HTML from our URL into the BeautifulSoup parse tree format
soup = BeautifulSoup(page, "lxml")

In [7]:
# Use either View Source command on your web page or the BeautifulSoup function 'prettify' to 
# look at the HTML underlying our chosen web page

print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgMonthNamesShort":["","Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"wgRequestId":"XlVG7QpAMFsAAL4CsBYAAADP","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":935851093,"wgRevisionId":935851093,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communi

In [8]:

# play around with some of the HTML tags and bring back the page 'title' and the data between the start and end 'title' tags
soup.title

# refine this a step further by specifying the 'string' element and only bring back the content without the 'title' tags
soup.title.string

'List of postal codes of Canada: M - Wikipedia'

### 'List of postal codes of Canada: M - Wikipedia'

In [11]:
# use the 'find_all' function to bring back all instances of the 'table' tag in the HTML and store in 'all_tables' variable
# all_tables=soup.find_all("table")
# all_tables

In [12]:
right_table=soup.find('table', class_='wikitable sortable')
right_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Harbourfront</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Lawrence_Heights" title="Lawrence Heights">Lawrence Heights</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North

In [13]:
A=[]
B=[]
C=[]

for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))

In [56]:
df=pd.DataFrame(A,columns=['Postcode'])
df['Borough']=B
df['Neighbourhood']=C
df

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
35       M7C      Not assigned   
36       M8C      Not assigned   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M2E      Not assigned   
45       M3E      Not assigned   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
49       M7E      Not assigned   
50       M8E      Not assigned   
51       M9E      Not assigned   
52       M1G       Scarborough   
53       M2G      Not assigned   
54       M3G      Not assigned   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
58       M7G      Not assigned   
59       M8G      Not assigned   
60       M9G      Not assigned   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
72       M7H      Not assigned   
73       M8H      Not assigned   
74       M9H      Not assigned   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
87       M7J      Not assigned   
88       M8J      Not assigned   
89       M9J      Not assigned   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M7K      Not assigned   
104      M8K      Not assigned   
105      M9K      Not assigned   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North 

### the 9th cell in the table on the Wikipedia page, the value of the Neighborhood columns will be Queen's Park.

In [45]:
df.iloc[[9], [2]] = "Queen's Park"
df

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
35       M7C      Not assigned   
36       M8C      Not assigned   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M2E      Not assigned   
45       M3E      Not assigned   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
49       M7E      Not assigned   
50       M8E      Not assigned   
51       M9E      Not assigned   
52       M1G       Scarborough   
53       M2G      Not assigned   
54       M3G      Not assigned   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
58       M7G      Not assigned   
59       M8G      Not assigned   
60       M9G      Not assigned   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
72       M7H      Not assigned   
73       M8H      Not assigned   
74       M9H      Not assigned   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
87       M7J      Not assigned   
88       M8J      Not assigned   
89       M9J      Not assigned   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M7K      Not assigned   
104      M8K      Not assigned   
105      M9K      Not assigned   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North 

In [46]:
df = df.replace('\n','', regex=True)
df

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
35       M7C      Not assigned   
36       M8C      Not assigned   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M2E      Not assigned   
45       M3E      Not assigned   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
49       M7E      Not assigned   
50       M8E      Not assigned   
51       M9E      Not assigned   
52       M1G       Scarborough   
53       M2G      Not assigned   
54       M3G      Not assigned   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
58       M7G      Not assigned   
59       M8G      Not assigned   
60       M9G      Not assigned   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
72       M7H      Not assigned   
73       M8H      Not assigned   
74       M9H      Not assigned   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
87       M7J      Not assigned   
88       M8J      Not assigned   
89       M9J      Not assigned   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M7K      Not assigned   
104      M8K      Not assigned   
105      M9K      Not assigned   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North 

In [47]:
df.shape

(287, 3)

In [48]:
df1 = df.sort_values(by=['Postcode'])
df1

Postcode           Borough  \
0        M1A      Not assigned   
11       M1B       Scarborough   
10       M1B       Scarborough   
28       M1C       Scarborough   
27       M1C       Scarborough   
26       M1C       Scarborough   
43       M1E       Scarborough   
42       M1E       Scarborough   
41       M1E       Scarborough   
52       M1G       Scarborough   
61       M1H       Scarborough   
75       M1J       Scarborough   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
108      M1L       Scarborough   
107      M1L       Scarborough   
106      M1L       Scarborough   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
139      M1N       Scarborough   
140      M1N       Scarborough   
152      M1P       Scarborough   
151      M1P       Scarborough   
150      M1P       Scarborough   
163      M1R       Scarborough   
164      M1R       Scarborough   
179      M1S       Scarborough   
191      M1T       Scarborough   
190      M1T       Scarborough   
192      M1T       Scarborough   
207      M1V       Scarborough   
205      M1V       Scarborough   
204      M1V       Scarborough   
206      M1V       Scarborough   
235      M1W       Scarborough   
245      M1X       Scarborough   
258      M1Y      Not assigned   
274      M1Z      Not assigned   
1        M2A      Not assigned   
12       M2B      Not assigned   
29       M2C      Not assigned   
44       M2E      Not assigned   
53       M2G      Not assigned   
62       M2H        North York   
78       M2J        North York   
77       M2J        North York   
76       M2J        North York   
93       M2K        North York   
109      M2L        North York   
110      M2L        North York   
125      M2M        North York   
126      M2M        North York   
141      M2N        North York   
153      M2P        North York   
165      M2R        North York   
180      M2S      Not assigned   
193      M2T      Not assigned   
208      M2V      Not assigned   
236      M2W      Not assigned   
246      M2X      Not assigned   
259      M2Y      Not assigned   
275      M2Z      Not assigned   
2        M3A        North York   
13       M3B        North York   
31       M3C        North York   
30       M3C        North York   
45       M3E      Not assigned   
54       M3G      Not assigned   
65       M3H        North York   
63       M3H        North York   
64       M3H        North York   
79       M3J        North York   
80       M3J        North York   
94       M3K        North York   
95       M3K        North York   
111      M3L        North York   
127      M3M        North York   
142      M3N        North York   
154      M3P      Not assigned   
166      M3R      Not assigned   
181      M3S      Not assigned   
194      M3T      Not assigned   
209      M3V      Not assigned   
237      M3W      Not assigned   
247      M3X      Not assigned   
260      M3Y      Not assigned   
276      M3Z      Not assigned   
3        M4A        North York   
15       M4B         East York   
14       M4B         East York   
32       M4C         East York   
46       M4E      East Toronto   
55       M4G         East York   
66       M4H         East York   
81       M4J         East York   
96       M4K      East Toronto   
97       M4K      East Toronto   
112      M4L      East Toronto   
113      M4L      East Toronto   
128      M4M      East Toronto   
143      M4N   Central Toronto   
155      M4P   Central Toronto   
167      M4R   Central Toronto   
182      M4S   Central Toronto   
196      M4T   Central Toronto   
195      M4T   Central Toronto   
213      M4V   Central Toronto   
212      M4V   Central Toronto   
211      M4V   Central Toronto   
210      M4V   Central Toronto   
214      M4V   Central Toronto   
238      M4W  Downtown Toronto   
248      M4X  Downtown Toronto   
249      M4X  Downtown Toronto   
261      M4Y  Downtown Tor

# 3. To create the above dataframe:

In [49]:
df_a = df[df.Borough != 'Not assigned']
df_a

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
52       M1G       Scarborough   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North York   
117      M6L        North York   
118      M6L        North York   
121      M9L        North York   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M2M        North York   
126      M2M        North York   
127      M3M        North York   
128      M4M      East Toronto   
129      M5M        North York   
130      M5M        North York   
131      M6M              York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
137      M9M        North York   
138      M9M        North York   
139      M1N       Scarborough   
140      M1N       Scarborough   
141      M2N        North York   
142      M3N        North York   
143      M4N   Central Toronto   
144      M5N   Central Toronto   
145      M6N              York   
146      M6N              York   
149      M9N              York   
150      M1P       Scarbor

In [50]:
df_a.shape

(210, 3)

In [21]:
df_b.shape

(210, 3)

## two rows will be combined into one row with the neighborhoods separated with a comma

In [193]:
df_b_2.sort_values(by=['Postcode'])

Postcode           Borough  \
10       M1B       Scarborough   
28       M1C       Scarborough   
41       M1E       Scarborough   
52       M1G       Scarborough   
61       M1H       Scarborough   
75       M1J       Scarborough   
90       M1K       Scarborough   
107      M1L       Scarborough   
122      M1M       Scarborough   
140      M1N       Scarborough   
152      M1P       Scarborough   
163      M1R       Scarborough   
179      M1S       Scarborough   
191      M1T       Scarborough   
207      M1V       Scarborough   
235      M1W       Scarborough   
245      M1X       Scarborough   
62       M2H        North York   
77       M2J        North York   
93       M2K        North York   
109      M2L        North York   
126      M2M        North York   
141      M2N        North York   
153      M2P        North York   
165      M2R        North York   
2        M3A        North York   
13       M3B        North York   
31       M3C        North York   
63       M3H        North York   
79       M3J        North York   
95       M3K        North York   
111      M3L        North York   
127      M3M        North York   
142      M3N        North York   
3        M4A        North York   
14       M4B         East York   
32       M4C         East York   
46       M4E      East Toronto   
55       M4G         East York   
66       M4H         East York   
81       M4J         East York   
97       M4K      East Toronto   
112      M4L      East Toronto   
128      M4M      East Toronto   
143      M4N   Central Toronto   
155      M4P   Central Toronto   
167      M4R   Central Toronto   
182      M4S   Central Toronto   
196      M4T   Central Toronto   
213      M4V   Central Toronto   
238      M4W  Downtown Toronto   
249      M4X  Downtown Toronto   
261      M4Y  Downtown Toronto   
4        M5A  Downtown Toronto   
17       M5B  Downtown Toronto   
33       M5C  Downtown Toronto   
47       M5E  Downtown Toronto   
56       M5G  Downtown Toronto   
67       M5H  Downtown Toronto   
82       M5J  Downtown Toronto   
98       M5K  Downtown Toronto   
114      M5L  Downtown Toronto   
129      M5M        North York   
144      M5N   Central Toronto   
156      M5P   Central Toronto   
170      M5R   Central Toronto   
184      M5S  Downtown Toronto   
199      M5T  Downtown Toronto   
219      M5V  Downtown Toronto   
239      M5W  Downtown Toronto   
251      M5X  Downtown Toronto   
5        M6A        North York   
18       M6B        North York   
34       M6C              York   
48       M6E              York   
57       M6G  Downtown Toronto   
71       M6H      West Toronto   
85       M6J      West Toronto   
101      M6K      West Toronto   
116      M6L        North York   
134      M6M              York   
145      M6N              York   
158      M6P      West Toronto   
172      M6R      West Toronto   
185      M6S      West Toronto   
7        M7A  Downtown Toronto   
173      M7R       Mississauga   
264      M7Y      East Toronto   
225      M8V         Etobicoke   
243      M8W         Etobicoke   
256      M8X         Etobicoke   
269      M8Y         Etobicoke   
281      M8Z         Etobicoke   
9        M9A      Queen's Park   
21       M9B         Etobicoke   
40       M9C         Etobicoke   
121      M9L        North York   
138      M9M        North York   
149      M9N              York   
162      M9P         Etobicoke   
177      M9R         Etobicoke   
227      M9V         Etobicoke   
244      M9W         Etobicoke   

                                         Neighbourhood  
10                                      Rouge, Malvern  
28              Port Union, Rouge Hill, Highland Creek  
41                   Guildwood, Morningside, West Hill  
52                                              Woburn  
61                                           Cedarbrae  
75                                      Rouge, Malvern  
90         East Birchmount Park, Ionview, Kennedy Park  
107        

In [194]:
df_b_2.shape

(103, 3)

# shade   (103, 3)

# 4. Submit a link to your Notebook on your Github repository.